### ISSUE TITLE
Incosistent Color Channel Order in examples/neural_style_transfer.py

### ISSUE THREAD

**Issue by dolaameng** [fchollet/keras#3340]
I notice there is a difference in how images are pre-processed in the [neural_style_transfer example](https://github.com/fchollet/keras/blob/master/examples/neural_style_transfer.py) and the [recommended way](https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3) of using VGG16 model. Specially,

- The color channels in neural_style_transfer example is RGB (as read by scipy.misc.imread), whereas the order in original VGG is expected to be BGR. 
-  The mean-centering step is not performed in the example, i.e.,
```
im[:,:,0] -= 103.939
im[:,:,1] -= 116.779
im[:,:,2] -= 123.68
```

Some of my experiments showed that these differences could result in sub-optimal results in image classification, e.g., [red fox](https://www.google.com.sg/?ion=1&espv=2#q=red%20fox) misclassified to [grey fox](https://www.google.com.sg/?ion=1&espv=2#q=grey%20fox) due to the wrong order of color channel. 

However, when it comes to generating "artistic" images as in the neural_style_transfer example, it seems that the results caused by the difference are not so different.  I can understand there are probably some reasons behind this,
- 1. The "content loss" and "total variation loss" defined in the example don't depend on the order of color channels.
- 2. The "style loss" might depend on the color orders, because it depends on the difference of "styles"  in the original and the generated image. And this style is measured by the inner-product distribution of different vgg filters (Gram matrix). However, it should be expected that most of vgg filters will be based on a mix of colors (e.g., edges, corners) rather than a single color alone (e.g, blue color for sea).

Those could be the reasons why I don't see many differences when experimenting with the different color channel encoding in the neural_style_transfer example. _But I really want to know whether we can come up with some contrived cases where the different color channel order will result in a difference for the neural_style_transfer example._

I post this question here because it might lead to a potential bug report to the above example code.

**Comment by titu1994** [comment]
Hello, this is my first time commenting here so please bear with any mistakes.

I have an implementation of this script, which simply exposes all the parameters of the script. It can be found at [titu1994/Neural-Style-Transfer](https://github.com/titu1994/Neural-Style-Transfer). The results here are based on the Network.py script there, which is a replica of the example and simply exposes the variables for modification.

I performed a few tests after implementing your suggestions. The result can be found [here](http://imgur.com/a/kdY7Z). 

A few points on the results : Both tests were done by converting RGB to BGR first, and then in one, mean centering was applied while in the other it was skipped.
-  As can be seen, with mean centering, the results are bad. Loss value fluctuates by decreasing in one iteration and increasing the next.
-  Without mean centering, almost same image is produced by the network as in the examples.

Therefore, I do not think converting RGB to BGR will offer much benefit, but maybe this image is not representative of the effects the change may cause. On the point of mean centering, I have performed several other tests and all produce the same image distortion.  

To duplicate the results shown here, please use the image in the content and style images folder (blue-moon-lake.jpg and starry_night.jpg) with the parameter flags as 
` --image_size 400 --content_weight 0.25 --style_weight 1 --total_variation_weight 1E-05 --style_scale 1 --num_iter 10 --rescale_image "False" --rescale_method "bilinear" --pool_type "ave"`

**Comment by dolaameng** [comment]
Hi @titu1994 thank fou for your comments. Yes I have observed similar things in my experiments.

1. The difference between RGB and BGR without mean centering is small. I have observed some cases where RGB tend to generate "purple" for "blue" parts in the original images, but was not sure if they were just coincidences .

2. The loss values fluctuated when mean centering was on. I have also initialized the generated images with values between -125 to 125 instead of 0 to 255. And the generated pixel values quickly went wild to a rang between -450 to 450 before clipping, whereas in the original implementation they were almost always in range -15 to 260. 

I haven't fingered out the reason behind this behavior, but my guess would be that different weights for these loss functions should be used after mean centeing, because it will change the ranges of them. I will try to post some results later if possible.

Again, I am a little bothered by the inconsistency in the keras example for vgg usage, specially considering a big community behind it and any small changes could be propagated. Or am I the one who has this issue? :)

**Comment by dolaameng** [comment]
I figured out why the loss values fluctuated in my implementation when the "BGR converting" and "mean centering" was on. It turned out to be a bug in my implementation of `deprocess_image()` - the changes were directly made to the image vector, including both compensation for the center-subtraction and clipping - so the loss values are decreased during the `l-bfgs-b` optimization, but will jump higher at the beginning of the next iteration because the `x` value is contaminated by `deprecoess_image()` 

A fix of the bug brings the performance back to the level of the original implementation - please see the [plot of loss values](http://imgur.com/a/zKsRI) comparing the two implementations - for the first 5 iterations durin g`l-bfgs-b`. And I observed that,
-  Content loss becomes lower, because of the halving-effect of `relu` to the negative values?
-  Style loss is lower, again probably because of the value ranges
-  Total Variation loss remains the same - as expected 
-  As a result, total loss starts with a lower value but the convergence rate is about the same.

At least now we know doing "color channel conversion" and "mean-centering" would NOT hurt the performance. Please see the modified example code [here](https://github.com/dolaameng/keras/blob/master/examples/pretrained_word_embeddings.py)

**Comment by titu1994** [comment]
@dolaameng It seems I have been making the exact same mistake in my implementations of the Network.py and INetwork.py. Thank you for the correction.

Not only does mean centering reduce the total loss, the final image produced is eerily similar to the standard image that is produced by the DeepArt.io algorithm, which utilizes Markov Random Field regularizers and Image Analogies algorithm to improve the results. Albeit similar, the current version still does not appear as good as the one produced by DeepArt.io.  Nevertheless, it is a step in the right direction.

For comparison, please see the images [here](http://imgur.com/a/EyVZu)
To be noted:
-  The average pooling method, while looking similar to DeepArt.io result, is far "smoother".
-  The max pooling method approaches near DeepArt.io result quality, without much local coherence.
-  Mean-centering produces images which are more visually appealing than the ones produced without, as can be seen when comparing the mean-centered version of these images to the examples provided.

As a result, I think we can conclude that the changes should be incorporated into the official neural_style_transfer.py script. I will be incorporating these changes to both the scripts Network.py and INetwork.py in my project since the results are compelling.

**Comment by dolaameng** [comment]
Hi @titu1994 It's great to hear that! Yes I will try to make a PR later. Thank you for your code and comments!

**Comment by dolaameng** [comment]
PR created at #3347
